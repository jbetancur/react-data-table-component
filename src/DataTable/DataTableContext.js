import React, { PureComponent, createContext } from 'react';
import PropTypes from 'prop-types';

export const DataTableStateContext = createContext();

export class DataTableProvider extends PureComponent {
  static propTypes = {
    initialState: PropTypes.object.isRequired,
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node,
    ]).isRequired,
  };

  render() {
    const { children, initialState } = this.props;

    return (
      <DataTableStateContext.Provider value={initialState}>
        {children}
      </DataTableStateContext.Provider>
    );
  }
}
