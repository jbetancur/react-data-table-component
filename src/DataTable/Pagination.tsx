import * as React from 'react';
import styled from 'styled-components';
import Select from './Select';
import { getNumberOfPages } from './util';
import useWindowSize from '../hooks/useWindowSize';
import useRTL from '../hooks/useRTL';
import { media, SMALL } from './media';
import { Direction } from './constants';
import { PaginationOptions } from './types';
import { defaultProps } from './defaultProps';

const defaultComponentOptions = {
	rowsPerPageText: 'Rows per page:',
	rangeSeparatorText: 'of',
	noRowsPerPage: false,
	selectAllRowsItem: false,
	selectAllRowsItemText: 'All',
};

const PaginationWrapper = styled.nav`
	display: flex;
	flex: 1 1 auto;
	justify-content: flex-end;
	align-items: center;
	box-sizing: border-box;
	padding-right: 8px;
	padding-left: 8px;
	width: 100%;
	${({ theme }) => theme.pagination.style};
`;

const Button = styled.button<{
	isRTL: boolean;
}>`
	position: relative;
	display: block;
	user-select: none;
	border: none;
	${({ theme }) => theme.pagination.pageButtonsStyle};
	${({ isRTL }) => isRTL && 'transform: scale(-1, -1)'};
`;

const PageList = styled.div`
	display: flex;
	align-items: center;
	border-radius: 4px;
	white-space: nowrap;
	${media.sm`
    width: 100%;
    justify-content: space-around;
  `};
`;

const Span = styled.span`
	flex-shrink: 1;
	user-select: none;
`;

const Range = styled(Span)`
	margin: 0 24px;
`;

const RowLabel = styled(Span)`
	margin: 0 4px;
`;

interface PaginationProps {
	rowsPerPage: number;
	rowCount: number;
	currentPage: number;
	direction?: Direction;
	paginationRowsPerPageOptions?: number[];
	paginationIconLastPage?: React.ReactNode;
	paginationIconFirstPage?: React.ReactNode;
	paginationIconNext?: React.ReactNode;
	paginationIconPrevious?: React.ReactNode;
	paginationComponentOptions?: PaginationOptions;
	onChangePage: (page: number) => void;
	onChangeRowsPerPage: (numRows: number, currentPage: number) => void;
}

function Pagination({
	rowsPerPage,
	rowCount,
	currentPage,
	direction = defaultProps.direction,
	paginationRowsPerPageOptions = defaultProps.paginationRowsPerPageOptions,
	paginationIconLastPage = defaultProps.paginationIconLastPage,
	paginationIconFirstPage = defaultProps.paginationIconFirstPage,
	paginationIconNext = defaultProps.paginationIconNext,
	paginationIconPrevious = defaultProps.paginationIconPrevious,
	paginationComponentOptions = defaultProps.paginationComponentOptions,
	onChangeRowsPerPage = defaultProps.onChangeRowsPerPage,
	onChangePage = defaultProps.onChangePage,
}: PaginationProps): JSX.Element {
	const windowSize = useWindowSize();
	const isRTL = useRTL(direction);
	const shouldShow = windowSize.width && windowSize.width > SMALL;
	// const isRTL = detectRTL(direction);
	const numPages = getNumberOfPages(rowCount, rowsPerPage);
	const lastIndex = currentPage * rowsPerPage;
	const firstIndex = lastIndex - rowsPerPage + 1;
	const disabledLesser = currentPage === 1;
	const disabledGreater = currentPage === numPages;
	const options = { ...defaultComponentOptions, ...paginationComponentOptions };
	const range =
		currentPage === numPages
			? `${firstIndex}-${rowCount} ${options.rangeSeparatorText} ${rowCount}`
			: `${firstIndex}-${lastIndex} ${options.rangeSeparatorText} ${rowCount}`;

	const handlePrevious = React.useCallback(() => onChangePage(currentPage - 1), [currentPage, onChangePage]);
	const handleNext = React.useCallback(() => onChangePage(currentPage + 1), [currentPage, onChangePage]);
	const handleFirst = React.useCallback(() => onChangePage(1), [onChangePage]);
	const handleLast = React.useCallback(
		() => onChangePage(getNumberOfPages(rowCount, rowsPerPage)),
		[onChangePage, rowCount, rowsPerPage],
	);
	const handleRowsPerPage = React.useCallback(
		(e: React.ChangeEvent<HTMLSelectElement>) => onChangeRowsPerPage(Number(e.target.value), currentPage),
		[currentPage, onChangeRowsPerPage],
	);

	const selectOptions = paginationRowsPerPageOptions.map((num: number) => (
		<option key={num} value={num}>
			{num}
		</option>
	));

	if (options.selectAllRowsItem) {
		selectOptions.push(
			<option key={-1} value={rowCount}>
				{options.selectAllRowsItemText}
			</option>,
		);
	}

	const select = (
		<Select onChange={handleRowsPerPage} defaultValue={rowsPerPage} aria-label={options.rowsPerPageText}>
			{selectOptions}
		</Select>
	);

	return (
		<PaginationWrapper className="rdt_Pagination">
			{!options.noRowsPerPage && shouldShow && (
				<>
					<RowLabel>{options.rowsPerPageText}</RowLabel>
					{select}
				</>
			)}
			{shouldShow && <Range>{range}</Range>}
			<PageList>
				<Button
					id="pagination-first-page"
					type="button"
					aria-label="First Page"
					aria-disabled={disabledLesser}
					onClick={handleFirst}
					disabled={disabledLesser}
					isRTL={isRTL}
				>
					{paginationIconFirstPage}
				</Button>

				<Button
					id="pagination-previous-page"
					type="button"
					aria-label="Previous Page"
					aria-disabled={disabledLesser}
					onClick={handlePrevious}
					disabled={disabledLesser}
					isRTL={isRTL}
				>
					{paginationIconPrevious}
				</Button>

				{!shouldShow && select}

				<Button
					id="pagination-next-page"
					type="button"
					aria-label="Next Page"
					aria-disabled={disabledGreater}
					onClick={handleNext}
					disabled={disabledGreater}
					isRTL={isRTL}
				>
					{paginationIconNext}
				</Button>

				<Button
					id="pagination-last-page"
					type="button"
					aria-label="Last Page"
					aria-disabled={disabledGreater}
					onClick={handleLast}
					disabled={disabledGreater}
					isRTL={isRTL}
				>
					{paginationIconLastPage}
				</Button>
			</PageList>
		</PaginationWrapper>
	);
}

export default React.memo(Pagination);
