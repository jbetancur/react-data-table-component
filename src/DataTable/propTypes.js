import React from 'react';
import PropTypes from 'prop-types';
import FirstPageIcon from '../icons/FirstPage';
import LastPageIcon from '../icons/LastPage';
import LeftIcon from '../icons/Left';
import RightIcon from '../icons/Right';
import ExpanderCollapsedIcon from '../icons/ExpanderCollapsedIcon';
import ExpanderExpandedIcon from '../icons/ExpanderExpandedIcon';

export const propTypes = {
  title: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
  ]),
  selectableRows: PropTypes.bool,
  expandableRows: PropTypes.bool,
  expandableDisabledField: PropTypes.string,
  defaultExpandedField: PropTypes.string,
  keyField: PropTypes.string,
  progressPending: PropTypes.bool,
  progressComponent: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  progressCentered: PropTypes.bool,
  expandableRowsComponent: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
    PropTypes.func,
  ]),
  expandableIcon: PropTypes.shape({
    collapsed: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.node,
      PropTypes.func,
    ]),
    expanded: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.node,
      PropTypes.func,
    ]),
  }),
  selectableRowsComponent: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  selectableRowsComponentProps: PropTypes.object,
  customTheme: PropTypes.object,
  sortIcon: PropTypes.oneOfType([PropTypes.bool, PropTypes.node]),
  sortFunction: PropTypes.func,
  sortServer: PropTypes.bool,
  onSort: PropTypes.func,
  striped: PropTypes.bool,
  highlightOnHover: PropTypes.bool,
  pointerOnHover: PropTypes.bool,
  actions: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  contextTitle: PropTypes.string,
  contextActions: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  onTableUpdate: PropTypes.func,
  onRowSelected: PropTypes.func,
  clearSelectedRows: PropTypes.bool,
  defaultSortField: PropTypes.string,
  defaultSortAsc: PropTypes.bool,
  columns: PropTypes.array,
  data: PropTypes.array,
  className: PropTypes.string,
  style: PropTypes.object,
  responsive: PropTypes.bool,
  overflowY: PropTypes.bool,
  overflowYOffset: PropTypes.string,
  noDataComponent: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  disabled: PropTypes.bool,
  noHeader: PropTypes.bool,
  subHeader: PropTypes.bool,
  subHeaderAlign: PropTypes.string,
  subHeaderWrap: PropTypes.bool,
  subHeaderComponent: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
    PropTypes.string,
  ]),
  onRowClicked: PropTypes.func,
  fixedHeader: PropTypes.bool,
  fixedHeaderScrollHeight: PropTypes.string,
  pagination: PropTypes.bool,
  paginationServer: PropTypes.bool,
  paginationDefaultPage: PropTypes.number,
  paginationTotalRows: PropTypes.number,
  paginationPerPage: PropTypes.number,
  paginationRowsPerPageOptions: PropTypes.array,
  onChangePage: PropTypes.func,
  onChangeRowsPerPage: PropTypes.func,
  paginationComponent: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
    PropTypes.func,
  ]),
  paginationComponentOptions: PropTypes.object,
  paginationIconFirstPage: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  paginationIconLastPage: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  paginationIconNext: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  paginationIconPrevious: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  selectableRowsDisabledField: PropTypes.string,
};

export const defaultProps = {
  title: '',
  keyField: 'id',
  selectableRows: false,
  expandableRows: false,
  expandableDisabledField: '',
  defaultExpandedField: '',
  progressPending: false,
  progressComponent: <h2>Loading...</h2>,
  progressCentered: false,
  expandableRowsComponent: <div>Add a custom expander component. Use props.data for row data</div>,
  expandableIcon: {
    collapsed: <ExpanderCollapsedIcon />,
    expanded: <ExpanderExpandedIcon />,
  },
  selectableRowsComponent: 'input',
  selectableRowsComponentProps: {},
  customTheme: {},
  sortIcon: false,
  sortFunction: null,
  sortServer: false,
  onSort: () => null,
  striped: false,
  highlightOnHover: false,
  pointerOnHover: false,
  contextTitle: '',
  contextActions: [],
  onTableUpdate: null,
  onRowSelected: () => null,
  clearSelectedRows: false,
  defaultSortField: null,
  defaultSortAsc: true,
  columns: [],
  data: [],
  className: null,
  style: {},
  responsive: true,
  overflowY: false,
  overflowYOffset: '250px',
  noDataComponent: 'There are no records to display',
  disabled: false,
  noHeader: false,
  subHeader: false,
  subHeaderAlign: 'right',
  subHeaderWrap: true,
  subHeaderComponent: [],
  onRowClicked: () => null,
  fixedHeader: false,
  fixedHeaderScrollHeight: '100vh',
  pagination: false,
  paginationServer: false,
  paginationDefaultPage: 1,
  paginationTotalRows: 0,
  paginationPerPage: 10,
  paginationRowsPerPageOptions: [10, 15, 20, 25, 30],
  onChangePage: () => null,
  onChangeRowsPerPage: () => null,
  paginationComponent: null,
  paginationComponentOptions: {
    rowsPerPageText: 'Rows per page:',
    rangeSeparatorText: 'of',
  },
  paginationIconFirstPage: <FirstPageIcon />,
  paginationIconLastPage: <LastPageIcon />,
  paginationIconNext: <RightIcon />,
  paginationIconPrevious: <LeftIcon />,
  selectableRowsDisabledField: '',
};
