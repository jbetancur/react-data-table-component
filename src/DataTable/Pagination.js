import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { useTableContext } from './DataTableContext';
import Select from './Select';
import { getNumberOfPages, detectRTL } from './util';
import useWindowSize from '../hooks/useWindowSize';
import { media, SMALL } from './media';

const defaultComponentOptions = {
  rowsPerPageText: 'Rows per page:',
  rangeSeparatorText: 'of',
  noRowsPerPage: false,
  selectAllRowsItem: false,
  selectAllRowsItemText: 'All',
};

const PaginationWrapper = styled.nav`
  display: flex;
  flex: 1 1 auto;
  justify-content: flex-end;
  align-items: center;
  box-sizing: border-box;
  padding-right: 8px;
  padding-left: 8px;
  width: 100%;
  ${props => props.theme.pagination.style};
`;

const Button = styled.button`
  position: relative;
  display: block;
  user-select: none;
  border: none;
  ${props => props.theme.pagination.pageButtonsStyle};
  ${props => props.isRTL && 'transform: scale(-1, -1)'};
`;

const PageList = styled.div`
  display: flex;
  align-items: center;
  border-radius: 4px;
  white-space: nowrap;
  ${media.sm`
    width: 100%;
    justify-content: space-around;
  `};
`;

const Span = styled.span`
  flex-shrink: 1;
  user-select: none;
`;

const Range = styled(Span)`
  margin: 0 24px;
`;

const RowLabel = styled(Span)`
  margin: 0 4px;
`;

const Pagination = ({
  rowsPerPage,
  rowCount,
  onChangePage,
  onChangeRowsPerPage,
  currentPage,
}) => {
  const {
    data,
    direction,
    paginationRowsPerPageOptions,
    paginationIconLastPage,
    paginationIconFirstPage,
    paginationIconNext,
    paginationIconPrevious,
    paginationComponentOptions,
  } = useTableContext();
  const windowSize = useWindowSize();
  const shouldShow = windowSize.width > SMALL;
  const isRTL = detectRTL(direction);
  const numPages = getNumberOfPages(rowCount, rowsPerPage);
  const lastIndex = currentPage * rowsPerPage;
  const firstIndex = (lastIndex - rowsPerPage) + 1;
  const disabledLesser = currentPage === 1;
  const disabledGreater = currentPage === numPages;
  const options = { ...defaultComponentOptions, ...paginationComponentOptions };
  const range = currentPage === numPages
    ? `${firstIndex}-${rowCount} ${options.rangeSeparatorText} ${rowCount}`
    : `${firstIndex}-${lastIndex} ${options.rangeSeparatorText} ${rowCount}`;

  const handlePrevious = useCallback(() => onChangePage(currentPage - 1), [currentPage, onChangePage]);
  const handleNext = useCallback(() => onChangePage(currentPage + 1), [currentPage, onChangePage]);
  const handleFirst = useCallback(() => onChangePage(1), [onChangePage]);
  const handleLast = useCallback(() => onChangePage(getNumberOfPages(rowCount, rowsPerPage)), [onChangePage, rowCount, rowsPerPage]);
  const handleRowsPerPage = useCallback(({ target }) => onChangeRowsPerPage(Number(target.value), currentPage), [currentPage, onChangeRowsPerPage]);

  const selectOptions = paginationRowsPerPageOptions.map(num => (
    <option
      key={num}
      value={num}
    >
      {num}
    </option>
  ));

  if (options.selectAllRowsItem) {
    selectOptions.push(
      (
        <option
          key={-1}
          value={data.length}
        >
          {options.selectAllRowsItemText}
        </option>
      ),
    );
  }

  const select = (
    <Select onChange={handleRowsPerPage} defaultValue={rowsPerPage}>
      {selectOptions}
    </Select>
  );

  return (
    <PaginationWrapper className="rdt_Pagination">
      {!options.noRowsPerPage && shouldShow && (
        <>
          <RowLabel>{options.rowsPerPageText}</RowLabel>
          {select}
        </>
      )}
      {shouldShow && (
        <Range>
          {range}
        </Range>
      )}
      <PageList>
        <Button
          id="pagination-first-page"
          type="button"
          aria-label="First Page"
          aria-disabled={disabledLesser}
          onClick={handleFirst}
          disabled={disabledLesser}
          isRTL={isRTL}
        >
          {paginationIconFirstPage}
        </Button>

        <Button
          id="pagination-previous-page"
          type="button"
          aria-label="Previous Page"
          aria-disabled={disabledLesser}
          onClick={handlePrevious}
          disabled={disabledLesser}
          isRTL={isRTL}
        >
          {paginationIconPrevious}
        </Button>

        {!shouldShow && select}

        <Button
          id="pagination-next-page"
          type="button"
          aria-label="Next Page"
          aria-disabled={disabledGreater}
          onClick={handleNext}
          disabled={disabledGreater}
          isRTL={isRTL}
        >
          {paginationIconNext}
        </Button>

        <Button
          id="pagination-last-page"
          type="button"
          aria-label="Last Page"
          aria-disabled={disabledGreater}
          onClick={handleLast}
          disabled={disabledGreater}
          isRTL={isRTL}
        >
          {paginationIconLastPage}
        </Button>
      </PageList>
    </PaginationWrapper>
  );
};

Pagination.propTypes = {
  rowsPerPage: PropTypes.number.isRequired,
  rowCount: PropTypes.number.isRequired,
  onChangePage: PropTypes.func.isRequired,
  onChangeRowsPerPage: PropTypes.func.isRequired,
  currentPage: PropTypes.number.isRequired,
};

export default Pagination;
