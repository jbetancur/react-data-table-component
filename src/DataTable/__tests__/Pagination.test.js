import 'jest-styled-components';
import React from 'react';
import { fireEvent } from '@testing-library/react';
import { renderWithTheme } from '../../test-helpers'; // since child elements require theme
import { DataTableProvider } from '../DataTableContext';
import { defaultProps } from '../propTypes';
import Pagination from '../Pagination';

const defaultStateMock = {
  ...defaultProps,
  paginationComponentOptions: {
    rowsPerPageText: 'Rows per page:',
    rangeSeparatorText: 'of',
  },
};

test('should render correctly with default props', () => {
  const { container } = renderWithTheme(
    <DataTableProvider initialState={{ ...defaultStateMock }}>
      <Pagination
        currentPage={1}
        rowsPerPage={10}
        rowCount={40}
        onChangePage={jest.fn()}
        onChangeRowsPerPage={jest.fn()}
      />
    </DataTableProvider>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

describe('when clicking the First Page button', () => {
  test('should call onChangePage with correct args when currentPage is > 1', () => {
    const onChangePageMock = jest.fn();
    const { container } = renderWithTheme(
      <DataTableProvider initialState={{ ...defaultStateMock }}>
        <Pagination
          currentPage={4}
          rowsPerPage={10}
          rowCount={40}
          onChangePage={onChangePageMock}
          onChangeRowsPerPage={jest.fn()}
        />
      </DataTableProvider>,
    );

    fireEvent.click(container.querySelector('button#pagination-first-page'));
    expect(onChangePageMock).toBeCalledWith(1);
  });

  test('should NOT call onChangePage with correct args when currentPage is === 1', () => {
    const onChangePageMock = jest.fn();
    const { container } = renderWithTheme(
      <DataTableProvider initialState={{ ...defaultStateMock }}>
        <Pagination
          currentPage={1}
          rowsPerPage={10}
          rowCount={40}
          onChangePage={onChangePageMock}
          onChangeRowsPerPage={jest.fn()}
        />
      </DataTableProvider>,
    );

    fireEvent.click(container.querySelector('button#pagination-first-page'));
    expect(onChangePageMock).not.toBeCalled();
  });
});

describe('when clicking the Last Page button', () => {
  test('should call onChangePage with correct args when currentPage is < the last page', () => {
    const onChangePageMock = jest.fn();
    const { container } = renderWithTheme(
      <DataTableProvider initialState={{ ...defaultStateMock }}>
        <Pagination
          currentPage={1}
          rowsPerPage={10}
          rowCount={40}
          onChangePage={onChangePageMock}
          onChangeRowsPerPage={jest.fn()}
        />
      </DataTableProvider>,
    );

    fireEvent.click(container.querySelector('button#pagination-last-page'));
    expect(onChangePageMock).toBeCalledWith(4);
  });

  test('should NOT call onChangePage with correct args when currentPage is the last page', () => {
    const onChangePageMock = jest.fn();
    const { container } = renderWithTheme(
      <DataTableProvider initialState={{ ...defaultStateMock }}>
        <Pagination
          currentPage={4}
          rowsPerPage={10}
          rowCount={40}
          onChangePage={onChangePageMock}
          onChangeRowsPerPage={jest.fn()}
        />
      </DataTableProvider>,
    );

    fireEvent.click(container.querySelector('button#pagination-last-page'));
    expect(onChangePageMock).not.toBeCalled();
  });
});

describe('when clicking the Next Page button', () => {
  test('should call onChangePage with correct args when currentPage is not the last page', () => {
    const onChangePageMock = jest.fn();
    const { container } = renderWithTheme(
      <DataTableProvider initialState={{ ...defaultStateMock }}>
        <Pagination
          currentPage={1}
          rowsPerPage={10}
          rowCount={40}
          onChangePage={onChangePageMock}
          onChangeRowsPerPage={jest.fn()}
        />
      </DataTableProvider>,
    );

    fireEvent.click(container.querySelector('button#pagination-next-page'));
    expect(onChangePageMock).toBeCalledWith(2);
  });

  test('should NOT call onChangePage with correct args when currentPage is the last page', () => {
    const onChangePageMock = jest.fn();
    const { container } = renderWithTheme(
      <DataTableProvider initialState={{ ...defaultStateMock }}>
        <Pagination
          currentPage={4}
          rowsPerPage={10}
          rowCount={40}
          onChangePage={onChangePageMock}
          onChangeRowsPerPage={jest.fn()}
        />
      </DataTableProvider>,
    );

    fireEvent.click(container.querySelector('button#pagination-next-page'));
    expect(onChangePageMock).not.toBeCalled();
  });
});

describe('when clicking the Previous Page button', () => {
  test('should call onChangePage with correct args when currentPage is > 1', () => {
    const onChangePageMock = jest.fn();
    const { container } = renderWithTheme(
      <DataTableProvider initialState={{ ...defaultStateMock }}>
        <Pagination
          currentPage={2}
          rowsPerPage={10}
          rowCount={40}
          onChangePage={onChangePageMock}
          onChangeRowsPerPage={jest.fn()}
        />
      </DataTableProvider>,
    );

    fireEvent.click(container.querySelector('button#pagination-previous-page'));
    expect(onChangePageMock).toBeCalledWith(1);
  });

  test('should NOT call onChangePage with correct args when currentPage is 1', () => {
    const onChangePageMock = jest.fn();
    const { container } = renderWithTheme(
      <DataTableProvider initialState={{ ...defaultStateMock }}>
        <Pagination
          currentPage={1}
          rowsPerPage={10}
          rowCount={40}
          onChangePage={onChangePageMock}
          onChangeRowsPerPage={jest.fn()}
        />
      </DataTableProvider>,
    );

    fireEvent.click(container.querySelector('button#pagination-previous-page'));
    expect(onChangePageMock).not.toBeCalled();
  });
});

describe('when there is no paging to be done', () => {
  test('should NOT call onChangePage with correct with any nav action when there are less rows that the page size', () => {
    const onChangePageMock = jest.fn();
    const { container } = renderWithTheme(
      <DataTableProvider initialState={{ ...defaultStateMock }}>
        <Pagination
          currentPage={1}
          rowsPerPage={10}
          rowCount={5}
          onChangePage={onChangePageMock}
          onChangeRowsPerPage={jest.fn()}
        />
      </DataTableProvider>,
    );

    fireEvent.click(container.querySelector('button#pagination-last-page'));
    expect(onChangePageMock).not.toBeCalled();

    onChangePageMock.mockReset();

    fireEvent.click(container.querySelector('button#pagination-first-page'));
    expect(onChangePageMock).not.toBeCalled();

    onChangePageMock.mockReset();

    fireEvent.click(container.querySelector('button#pagination-previous-page'));
    expect(onChangePageMock).not.toBeCalled();

    onChangePageMock.mockReset();

    fireEvent.click(container.querySelector('button#pagination-next-page'));
    expect(onChangePageMock).not.toBeCalled();
  });
});

describe('when clicking the Previous Page button', () => {
  test('should call onChangePage with correct args when currentPage is > 1', () => {
    const onChangeRowsPerPageMock = jest.fn();
    // TOOO: remove when trailing empty rows is fixed
    const onChangePageMock = jest.fn();
    const { container } = renderWithTheme(
      <DataTableProvider initialState={{ ...defaultStateMock }}>
        <Pagination
          currentPage={2}
          rowsPerPage={10}
          rowCount={40}
          onChangePage={onChangePageMock}
          onChangeRowsPerPage={onChangeRowsPerPageMock}
        />
      </DataTableProvider>,
    );

    fireEvent.change(container.querySelector('select'), { target: { value: 20 } });
    expect(onChangeRowsPerPageMock).toBeCalledWith(20, 2);
  });
});
