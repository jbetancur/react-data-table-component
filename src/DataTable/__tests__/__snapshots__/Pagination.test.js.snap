// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly with default props 1`] = `
.c0 {
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  font-size: 13px;
  color: rgba(0,0,0,.54);
}

<span
  class="c0"
>
  :
</span>
`;
